### DISCUSSION TITLE
Cloud-init: change default network interface name from eth0 to ens18

### DISCUSSION THREAD

**Discussion by everythings-gonna-be-alright** [bpg/terraform-provider-proxmox#847] [Q&A] [answered]
```
# ls -lh
total 2.0K
-rw-r--r-- 1 root root  54 Dec 23 23:24 meta-data
-rw-r--r-- 1 root root 313 Dec 23 23:24 network-config
-rw-r--r-- 1 root root 217 Dec 23 23:24 user-data
-rw-r--r-- 1 root root 276 Dec 23 23:24 vendor-data

# cat network-config | grep 'name:'
      name: eth0
```

How do I change the network interface name from `eth0` to `ens18`?
I migrated infrastructure from another provider, which used `ens18` name as default.
I run Kubernetes on VMs, and for the CNI plugin, having the same interface name on each compute node is vital.

**Comment by bpg-dev** [comment]
Hi @everythings-gonna-be-alright!

>How do I change the network interface name from eth0 to ens18?

The provider doesn't do anything with regards to interface names, and honestly I don't have much ideas where the name is coming from. This behaviour could be specific for a distro you're using, or something that Proxmox is adding on top of its auto-generate cloud-init configuration, which could also be specific to PVE version.
 
In theory, any customization can be done via a custom cloud-init config that you can provide via [`network_data_file_id`](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#network_data_file_id) attribute.

>I migrated infrastructure from another provider, which used ens18 name as default.

I don't have any context about "another provider", so can't comment on that.

> I run Kubernetes on VMs, and for the CNI plugin, having the same interface name on each compute node is vital.

Sorry, don't quite understand the problem. Wouldn't `eth0` name be same on all the nodes?

**Reply by vacquah** [reply]
Is there such a thing as :  ` network_interface {name = ""}`

I am using Talos Kubernetes, which seems to use eth0, but my pfsense bridge is on another network interface, not eth0,  so running into connectivity issues. I came here looking to do the same thing as the OP - set/change network interface name . Possible?

**Reply by bpg-dev** [reply]
Can this be done in the Proxmox UI? It seems more of a OS-level setting (or cloud-init customization) to me, rather than provider's feature.

**Comment by everythings-gonna-be-alright** [comment, marked as answer]
My bad. I thought generating `network-config` cloud-init template happens on the terraform provider side.
But actually, this provider sets the IP address and Gateway into `network-config` through Proxmox API. 
Moreover, Proxmox API doesn't provide the ability to set custom interface names to cloud-init network templates.
The best way to resolve this issue is to generate a custom configuration and override the Proxmox cloud-init network configuration ( `network_data_file_id` ). 
It is also necessary to manually generate mac_address and put it into the `network-config` template  ( if you want to set a custom interface name )
Or 
Just keep the default OS interface name. This way is a little bit dirty, but it will work. Here is an example:
```
version: 1
config:
    - type: physical
      name: ens18
      subnets:
      - type: static
        address: '${IP_ADDRESS}'
        netmask: '255.255.255.0'
        gateway: '10.15.0.1'
    - type: nameserver
      address:
      - '10.15.0.1'
      search:
      - 'my.com'
```
@bpg, thanks for your time.

**Comment by konstantin-kornienko** [comment]
Just sharing another possible solution.

in our case using `network_data_file_id` resource is not an option, because it requires ssh access to the proxmox node.

Instead, we had to rename an interface from `ens18` to `eth0` in our VM templates :(
This link can be used: https://techviewleo.com/change-network-interface-name-to-eth0-on-debian/
Btw the thing that names interfaces like `ens18` is called [Systemd predictable interface names](https://systemd.io/PREDICTABLE_INTERFACE_NAMES/). 

Thanks to @everythings-gonna-be-alright and @bpg for sharing the information!
